### PR TITLE
fix(docker): add cgroup driver setting, but do not overwrite deamon.json file

### PR DESCRIPTION
**Pull request by dmpe**
attempt at fixing https://github.com/kubernetes/minikube/issues/16323

## Logic:

- `/etc/docker/daemon.json` is attempted to be read
- if it does not exist, then file is created, with `native.cgroupdriver` key
- if it exists, then it is parsed. 
  - if such JSON does not containt `exec-opts`, then it is added (with only this key/value and cgroup as passed to the function)
  - if such JSSON contains already `exec-opts`, then it's value is checked.
    - if the key is `native.cgroupdriver` then we skip any further processing, and file is not touched anymore
    - if the key is not `native.cgroupdriver`, then we add it with the value of the parameter that was passed. No other keys such as e.g. `bip` are adjusted/overwritten.  

## Tests

Besides CI, also run:

```
# example 1
sudo ./out/minikube stop && sudo ./out/minikube delete
sudo ./out/minikube start --driver none --cni calico --force-systemd=true
```

with `/etc/docker/daemon.json` being one of many combinations:

```
{"bip":"192.168.1.5/24","exec-opts":["native.cgroupdriver=systemd"],"log-driver":"json-file","log-opts":{"max-size":"100m"},"storage-driver":"overlay2"}
```
```
{"bip":"192.168.1.5/24","log-driver":"json-file","log-opts":{"max-size":"100m"},"storage-driver":"overlay2"}
```

Locally I tested this on Ubuntu 23.04 + docker 23.0.5

### CODE REVIEW

**Comment by dmpe** [comment]
Hi reviewers, can you take a look please on this. If the logic/approach to it fine, could you pls trigger additional tests? Thanks

**Comment by dmpe** [comment]
Do you mind @spowelljr taking a look on this? I havent rebased for some time, but would appreciate a feedback - maybe even running some additional CI tests?

**Comment by medyagh** [comment]
@dmpe do you mind rebasing your PR, thanks for patience on the review

**Comment by dmpe** [comment]
> @dmpe do you mind rebasing your PR, thanks for patience on the review

Hi @medyagh, thank you very much for the feedback. I believe i have rebased the PR, shortly before you got back to me.

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:582]:
```suggestion
				return nil
			}
		}
		klog.Infof("Insert %s driver into %s file.", natCgroupDriver, DockerDaemonJSON)
		daemonData["exec-opts"] = append(jsonExecOpts, cgroupKeyValue)
		writeToDockerdJSON = true
```
If Cgroup is already expect can just return and save some nesting.

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:589]:
```suggestion
		return writeToDockerdConfig(daemonStr)
	}

	// we don't know exact structure that user has already been using
	err = json.Unmarshal(file, &daemonData)
	if err != nil {
		klog.Warningf("Failed to parse %s", file)
	}
	execOpts := daemonData["exec-opts"]
	if execOpts == nil {
		klog.Infof("%s does not contain any 'exec-opts' key. Configuring docker to use %s as cgroup driver...", DockerDaemonJSON, driver)
		daemonData["exec-opts"] = []string{cgroupKeyValue}
		return writeToDockerdConfig(daemonData)
	}
	klog.Infof("%s contains exec-opts option. Checking for %s key presence.", DockerDaemonJSON, natCgroupDriver)
	jsonExecOpts := execOpts.([]interface{})
	for _, value := range jsonExecOpts {
		dKey := strings.Split(value.(string), "=")[0]
		if dKey == natCgroupDriver {
			return nil
		}
	}
	klog.Infof("Insert %s driver into %s file.", natCgroupDriver, DockerDaemonJSON)
	daemonData["exec-opts"] = append(jsonExecOpts, cgroupKeyValue)
	return writeToDockerdConfig(daemonData)
}

func writeToDockerdConfig(data any) error {
	dataBytes, _ := json.MarshalIndent(data, "", "  ")
	ma := assets.NewMemoryAsset(dataBytes, "/etc/docker", "daemon.json", "0644")
	return r.Runner.Copy(ma)
```
Having a `writeToDockerdConfig` func would remove most of the nesting and make it a lot easier to read

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:544]:
```suggestion
	var daemonData map[string]any
```

**Review by spowelljr — CHANGES_REQUESTED** [review]

File: pkg/minikube/cruntime/docker.go
Comment (dmpe) [reply pkg/minikube/cruntime/docker.go]: Thanks @spowelljr ! I have incorporated the fixes. The `writeToDockerdConfig` had to be slightly adjusted as `(r *Docker)` struct type was missing.

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:565]:
```suggestion
	jsonExecOpts, ok := execOpts.([]interface{})
	if !ok {
```
Just incase the user malformed their config file (string instead of []string) we should use `ok` to prevent a panic.

**Review by spowelljr — CHANGES_REQUESTED** [review]

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:567]: Same with `value.(string)` if an `int` snuck it would cause a panic.

**Review by spowelljr — CHANGES_REQUESTED** [review]

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:565]:
```suggestion
	jsonExecOpts, ok := execOpts.([]string)
```
I believe that would prevent needing the `value.(string)` below.

**Comment by dmpe** [comment]
Thanks again.

File: pkg/minikube/cruntime/docker.go
Comment (spowelljr) [inline pkg/minikube/cruntime/docker.go:550]: If there's no file could you continue to use the JSON string we were previously writing. For non-none driver usage we're basically using this command to write the entire daemon.json but now it's only getting populated with the exec-opts, so continue writing the whole file to prevent unexpected issues.

**Before:**
```
{
"exec-opts": ["native.cgroupdriver=cgroupfs"],
"log-driver": "json-file",
"log-opts": {
	"max-size": "100m"
},
"storage-driver": "overlay2"
}
```

**Now:**
```
{
"exec-opts": ["native.cgroupdriver=cgroupfs"]
}
```

**Review by spowelljr — CHANGES_REQUESTED** [review]

**Comment by dmpe** [comment]
The `pull-minikube-build` now fails with `! Image was not built for the current minikube version. To resolve this you can delete and recreate your minikube cluster using the latest images. Expected minikube version: v1.31.0 -> Actual minikube version: v1.30.1`. 
In other news, I think i have adjusted as proposed.

**Comment by spowelljr** [comment]
> The `pull-minikube-build` now fails with `! Image was not built for the current minikube version. To resolve this you can delete and recreate your minikube cluster using the latest images. Expected minikube version: v1.31.0 -> Actual minikube version: v1.30.1`. In other news, I think i have adjusted as proposed.

That's fine, that's just because you rebased and we're about to go for a release to the kicbase is updated to expect the minikube version to be v1.31.0, you can ignore it

File: pkg/minikube/cruntime/docker.go
Comment (dmpe) [inline pkg/minikube/cruntime/docker.go:544]:
```suggestion
			"exec-opts": ["%s"],
```

File: pkg/minikube/cruntime/docker.go
Comment (dmpe) [inline pkg/minikube/cruntime/docker.go:550]:
```suggestion
		}`, cgroupKeyValue)
```

**Comment by dmpe** [comment]
I am not sure why a lot of tests are failing, and I have a _feeling_ it is not due to being flake, rather some code issue. E.g. [this test](https://storage.googleapis.com/minikube-builds/logs/16363/30224/KVM_Linux.html#fail_TestMultiNode/serial/StartAfterStop) failed when using cgroupfs (minikube start was without force systemd option).

**Comment by dmpe** [comment]
@spowelljr Hi, could you take a look on this again, sometime. Thanks

**Comment by dmpe** [comment]
Hi, Can you take a look and if it is possible merge this already ? @spowelljr @medyagh. Thanks.

File: pkg/minikube/cruntime/docker.go
Comment (dmpe) [reply pkg/minikube/cruntime/docker.go]: Has been addressed.

**Comment by dmpe** [comment]
> Hi, Can you take a look and if it is possible merge this already ? @spowelljr @medyagh. Thanks.

How I help to get this over the finishing line ? Am I missing something ?

The issue really prevents us from upgrading minikube to newer versions, because we require that some docker deamon options are set manually - not by minikube (e.g. `bip`). There are already other people impacted by it as well https://github.com/NVIDIA/k8s-device-plugin/issues/423 - I am however without the ability to switch to docker driver. 

Maybe @prezha you could take a look as well ? Many thanks.

**Comment by prezha** [comment]
hey @dmpe, looks like the underlying code changed in the meantime:

```
This branch has conflicts that must be resolved
Use the [web editor](https://github.com/kubernetes/minikube/pull/16363/conflicts) or the to resolve conflicts.
Conflicting files
pkg/minikube/cruntime/docker.go
```
also, users reported that the nvidia gpu addon works for them (eg, [this](https://github.com/NVIDIA/k8s-device-plugin/issues/423#issuecomment-1701735916) message)

can you please re-check?

**Comment by dmpe** [comment]
> hey @dmpe, looks like the underlying code changed in the meantime:
> 
> ```
> This branch has conflicts that must be resolved
> Use the [web editor](https://github.com/kubernetes/minikube/pull/16363/conflicts) or the to resolve conflicts.
> Conflicting files
> pkg/minikube/cruntime/docker.go
> ```
> 
> also, users reported that the nvidia gpu addon works for them (eg, [this](https://github.com/NVIDIA/k8s-device-plugin/issues/423#issuecomment-1701735916) message)
> 
> can you please re-check?

I have just rebased and hope that it is going to be fine, with all of you. Please, dont forget to approve the workflows else cannot be merged -:). cc @spowelljr 

Speaking of the nvidia issue I linked: The user has indeed "fixed" that by using a different driver - which however I am not able  to do (existing scripting, testing need for new approach, etc.). As I wrote previously, we require several additional options being set in daemon json file ... Hence overwriting this file by minikube installation is something that prohibits us from upgrading to the newer minikube versions for now. 

Many thanks.

File: pkg/minikube/cruntime/docker.go
Comment (prezha) [inline pkg/minikube/cruntime/docker.go:592]: nit: with a single return error value, we can compress it to the `if err := ...; err != nil {...` form

File: pkg/minikube/cruntime/docker.go
Comment (prezha) [inline pkg/minikube/cruntime/docker.go:593]: should it fail here, there's probably no point in continuing, as we might get some weird, unexpected results?

File: pkg/minikube/cruntime/docker.go
Comment (prezha) [inline pkg/minikube/cruntime/docker.go:599]: we should not ignore eventual error returned

File: pkg/minikube/cruntime/docker.go
Comment (prezha) [inline pkg/minikube/cruntime/docker.go:600]: this would only add missing `exec-opts` in the existing daemon.json and ignore/break the `nvidia-docker` config, as that would be discarded

File: pkg/minikube/cruntime/docker.go
Comment (prezha) [inline pkg/minikube/cruntime/docker.go:608]: similarly, this would ignore/break the nvidia-docker config, as that would be discarded

File: pkg/minikube/cruntime/docker.go
Comment (prezha) [inline pkg/minikube/cruntime/docker.go:619]: as above, this would ignore/break the nvidia-docker config, as that would be discarded

**Review by prezha — CHANGES_REQUESTED** [review]
@dmpe i left a few comments, please have a look
most critically is to more specifically account for and retain the "nvidia-docker" config, if applicable
you can log the config that'd be applied, and then please share _before_ (original) and _after_ (final) config outputs
ideally, we'd have tests to cover these

one thing i'm thinking of:
we're fine with overwriting our own `kicbase`/`iso` isolated docker daemon.json
the problem is only with the `none` driver - ie, when we're on a "bare-metal" or directly on a user's vm
should we, in the `none` driver case, **_just not touch the docker daemon.json_** and let users configure it appropriately (and also customise it to their needs)?
this would then also leave it to the user to configure the docker daemon for "nvidia-docker" (which might not be bad, as it's not hard-coded and is under users' control, should the config need to be changed in the future)
apart from this, historically, we needed to rewrite docker daemon.json just to set the correct cgroups driver (in our kicbase/iso), which should be automatically and correctly set by docker for `none` machines anyway

@medyagh @spowelljr @afbjorklund wdyt?

**Comment by dmpe** [comment]
Thank you again @prezha for both feedback and quick responses!

>  and let users configure it appropriately (and also customise it to their needs)

I _absolute_ AGREE - not just agree but also think it should have been done from day 1. And I mean, it has worked well before...

As a `none` user, you are already telling me at least twice that if I run minikube I need `cri-docker` these days and install `crictl` as well. Myself! Minikube does not do it, even though I certainly could ...

Your proposal would also fix the nvidia issue as well my additional configuration that I have to have. See the parent issue how my docker json file looks like.
Meaning that either I had to include additional docker options into the `struct` (which I considered a lot; but how many should minikube additionally expose or even include - all/some ?!) or use a different approach of merging default config with any additional one like here.

I would be **very happy** if this existing behaviour got changed, so that daemon json is fully in my own ownership. Minikube project could then tell me that I have to configure it properly myself and project can't make sure that it will work. And I am fine with that because you are already assuming that I am "expert user". 

I will wait for others to reply before going through you comments.

**Comment by dmpe** [comment]
@prezha how would I go about implementing what we actually agree on, namely let users configure their docker daemon config by themselves. 

There are at least 2 ways AFAIK:
- remove `force-systemd` parameter from `minikube start`
  - more code adjustments required (not bad!) but I am not sure if `non-systemd` driver is still being used somewhere - as of now, with 1.27+ being minikube's default k8s version. 
    -  e.g. `cgroupDriver` function will be removed, etc.
- only adjust `configureDocker` function to do "nothing" when `--driver` is set to `none` ?
  - less invasive to minikube code base because we only add if/else check.  

[Docker Documentation states following](https://docs.docker.com/engine/reference/commandline/dockerd/):
> If you omit the native.cgroupdriver option, cgroupfs is used on cgroup v1 hosts, systemd is used on cgroup v2 hosts with systemd available.

Considering that we are doing a new major release, and `systemd` become standard I would be in favor of entirely removing `cgroup` option(s) from minikube and rely on OS default/what docker/user sets by themselves. WDYT?

**Pull request closed without merging**
